api/lib/flight_job_script_api/subprocess.rb

```diff
@@ -0,0 +1,191 @@
+#==============================================================================
+# Copyright (C) 2021-present Alces Flight Ltd.
+#
+# This file is part of Flight Job Script Service.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# Flight Job Script Service is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with Flight Job Script Service. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on Flight Job Script Service, please visit:
+# https://github.com/openflighthpc/flight-job-script-service
+#==============================================================================
+
+module FlightJobScriptAPI
+  class Subprocess
+    class Result < Struct.new(:stdout, :stderr, :exitstatus, :pid); end
+
+    def initialize(dir:, env:, gid:, logger:, timeout:, uid:)
+      @dir = dir
+      @env = env
+      @gid = gid
+      @logger = logger
+      @timeout = timeout
+      @uid = uid
+    end
+
+    def run(cmd, stdin, &block)
+      @stdout = ""
+      @stderr = ""
+      @read_threads = []
+      start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+      create_pipes
+      run_fork do
+        run_command(cmd, &block)
+      end
+      start_read_threads
+      write_stdin(stdin)
+      status = wait_for_process
+      wait_for_read_threads(start_time)
+      exit_code = determine_exit_code(status)
+
+      Result.new(@stdout, @stderr, exit_code, @pid)
+    ensure
+      close_pipes
+    end
+
+    private
+
+    def run_fork(&block)
+      @pid = Kernel.fork do
+        @out_read.close
+        @err_read.close
+        @in_write.close
+
+        Process::Sys.setgid(@gid)
```

Comment:
IIRC we need to set the `groups` of the process as well as the `gid`? I vaguely remember seeing this in one of our applications.

Reply:
OK. You've convinced me. https://github.com/openflighthpc/flight-job-script-service/pull/41/commits/8dce3fdf21044634364661fff37985b787d45a72